featuretools/computational_backends/pandas_backend.py

```diff
@@ -233,6 +141,131 @@ def calculate_all_features(self, instance_ids, time_last,
+                                    columns=self.feature_set.necessary_columns[path],
+                                    time_last=time_last,
+                                    training_window=training_window)
+
+        # If the last edge was backward, copy the parent's link variables to
+        # this entity's dataframe.
+        if path:
+            is_forward, relationship = path[-1]
+            if not is_forward:
+                df, link_variables = self._add_link_vars(df, parent_df, relationship,
+                                                         parent_link_variables)
+
+                link_variables.append(relationship.child_variable.id)
+            else:
+                link_variables = []
+        else:
+            link_variables = []
+
+        # Recurse on children.
+        for edge, sub_trie in feature_trie.children():
+            is_forward, relationship = edge
+            if is_forward:
+                sub_entity = relationship.parent_entity
+                sub_filter_variable = relationship.parent_variable.id
+                sub_filter_values = df[relationship.child_variable.id]
+            else:
+                sub_entity = relationship.child_entity
+                sub_filter_variable = relationship.child_variable.id
+                sub_filter_values = df[relationship.parent_variable.id]
+
+            sub_df_trie = df_trie.get_node([edge])
+            self._calculate_features_on_trie(entity=sub_entity,
+                                             feature_trie=sub_trie,
+                                             df_trie=sub_df_trie,
+                                             path=path + [edge],
+                                             filter_variable=sub_filter_variable,
+                                             filter_values=sub_filter_values,
+                                             time_last=time_last,
+                                             training_window=training_window,
+                                             precalculated_features=precalculated_features,
+                                             parent_link_variables=link_variables,
+                                             parent_df=df,
+                                             ignored=ignored)
+
+        # Add any precalculated features.
+        if entity.id in precalculated_features:
+            precalc_feature_values = precalculated_features[entity.id]
+            # Left outer merge to keep all rows of df.
+            df = df.merge(precalc_feature_values,
+                          how='left',
+                          left_index=True,
+                          right_index=True,
+                          suffixes=('', '_precalculated'))
+
+        df_trie[[]] = df
+        feature_names = feature_trie[[]]
+        if ignored:
+            feature_names -= ignored
+        feature_groups = self.feature_set.group_features(feature_names)
+        for group in feature_groups:
```

Comment:
My intention was for `_calculate_feature_group` to only read from `df_trie`, return an updated dataframe, and then have this function update the trie with the new dataframe. I hoped that this would make it clear when `df_trie` is updated, but it may have just added complexity.

Two other options:

1. Make each feature type handler responsible for updating `df_trie`. This loop would then just be:
```py
for group in feature_groups:
    self._calculate_feature_group(group, df_trie)

# Now df_trie[[]] has all features calculated
```
2. Put the dataframe for this entity into `df_trie` only at the end of `_calculate_features_on_trie`. We would then need to pass `df` to `_calculate_feature_groups` so that the type handlers can access it (instead of accessing it from `df_trie`):
```py
for group in feature_groups:
    df = self._calculate_feature_group(group, df, df_trie)

df_trie[[]] = df
```

Reply:
I get what's going on now and think it's reasonable. I think a few comments might help people like me who look at the code too quickly. 

what i missed is 
1. Every feature in a group is operating on the same entity
2. You are storing the new df in `df_trie`. Then `df_trie` is getting passed back into the next calculation which is only going to append columns to the existing df, so you can resave it in the trie with impacting the previous calculation.

Reply:
it's possible that removing the `_calculate_feature_group` method it might help. it really isn't doing much and is only called here